Compiler crash after building the project for the second time after installing ballerina as root in ubuntu

After installing the ballerina following Linux instructions as the root (Ubuntu 20.04.2 LTS), compiler crashes when building a project without root access for the second time.

Ballerina version : Ballerina Swan Lake Alpha 4

ballerina-internal.log output is shown below
`[2021-04-19 11:51:25,167] SEVERE {b7a.log.crash} - Failed to write library: /home/damitha/Documents/balHello/helloworld/target/cache/damitha/helloworld/0.1.0/java11/helloworld.connection.jar 
java.lang.RuntimeException: Failed to write library: /home/damitha/Documents/balHello/helloworld/target/cache/damitha/helloworld/0.1.0/java11/helloworld.connection.jar
	at io.ballerina.projects.repos.FileSystemCache.cachePlatformSpecificLibrary(FileSystemCache.java:114)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:271)
	at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:399)
	at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:327)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:145)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:131)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:106)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:144)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:105)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:69)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:251)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
Caused by: java.io.IOException: File '/home/damitha/Documents/balHello/helloworld/target/cache/damitha/helloworld/0.1.0/java11/helloworld.connection.jar' cannot be written to
	at org.apache.commons.io.FileUtils.openOutputStream(FileUtils.java:355)
	at org.apache.commons.io.FileUtils.writeByteArrayToFile(FileUtils.java:2222)
	at org.apache.commons.io.FileUtils.writeByteArrayToFile(FileUtils.java:2186)
	at org.apache.commons.io.FileUtils.writeByteArrayToFile(FileUtils.java:2171)
	at io.ballerina.projects.repos.FileSystemCache.cachePlatformSpecificLibrary(FileSystemCache.java:111)`

--- a/cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java
+++ b/cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java
@@ -202,7 +202,7 @@ public class BuildCommand implements BLauncherCmd {
             try {
                 project = SingleFileProject.load(this.projectPath, buildOptions);
             } catch (ProjectException e) {
-                CommandUtil.printError(this.errStream, e.getMessage(), buildCmd, false);
+                CommandUtil.printError(this.errStream, e.getMessage(), null, false);
                 CommandUtil.exitError(this.exitWhenFinish);
                 return;
             }
@@ -221,7 +221,7 @@ public class BuildCommand implements BLauncherCmd {
             try {
                 project = BuildProject.load(this.projectPath, buildOptions);
             } catch (ProjectException e) {
-                CommandUtil.printError(this.errStream, e.getMessage(), buildCmd, false);
+                CommandUtil.printError(this.errStream, e.getMessage(), null, false);
                 CommandUtil.exitError(this.exitWhenFinish);
                 return;
             }

--- a/cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CleanTargetDirTask.java
+++ b/cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CleanTargetDirTask.java
@@ -36,7 +36,6 @@ public class CleanTargetDirTask implements Task {
     @Override
     public void execute(Project project) {
         try {
-            ProjectUtils.checkWritePermission(project.sourceRoot());
             Target target = new Target(project.sourceRoot());
             target.clean();
         } catch (IOException | ProjectException e) {

--- a/cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CleanTargetDirTask.java
+++ b/cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CleanTargetDirTask.java
@@ -21,7 +21,6 @@ package io.ballerina.cli.task;
 import io.ballerina.projects.Project;
 import io.ballerina.projects.ProjectException;
 import io.ballerina.projects.internal.model.Target;
-import io.ballerina.projects.util.ProjectUtils;
 
 import java.io.IOException;
 

--- a/cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java
+++ b/cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java
@@ -40,7 +40,6 @@ import static io.ballerina.cli.launcher.LauncherUtils.createLauncherException;
 import static io.ballerina.cli.utils.FileUtils.getFileNameWithoutExtension;
 import static io.ballerina.projects.util.ProjectConstants.BLANG_COMPILED_JAR_EXT;
 import static io.ballerina.projects.util.ProjectConstants.USER_DIR;
-import static io.ballerina.projects.util.ProjectUtils.checkWritePermission;
 
 /**
  * Task for creating the executable jar file.
@@ -68,16 +67,6 @@ public class CreateExecutableTask implements Task {
         Target target;
 
         try {
-            if (project.kind().equals(ProjectKind.SINGLE_FILE_PROJECT)) {
-                checkWritePermission(currentDir);
-            } else {
-                checkWritePermission(project.sourceRoot());
-            }
-        } catch (ProjectException e) {
-            throw createLauncherException(e.getMessage());
-        }
-
-        try {
             if (project.kind().equals(ProjectKind.BUILD_PROJECT)) {
                 target = new Target(project.sourceRoot());
             } else {
@@ -86,6 +75,8 @@ public class CreateExecutableTask implements Task {
             }
         } catch (IOException e) {
             throw createLauncherException("unable to resolve target path:" + e.getMessage());
+        } catch (ProjectException e) {
+            throw createLauncherException("unable to create executable:" + e.getMessage());
         }
 
         Path executablePath;

--- a/compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java
+++ b/compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java
@@ -70,7 +70,6 @@ import java.util.jar.Manifest;
 import java.util.stream.Collectors;
 
 import static io.ballerina.projects.util.FileUtils.getFileNameWithoutExtension;
-import static io.ballerina.projects.util.ProjectUtils.checkWritePermission;
 import static org.ballerinalang.compiler.CompilerOptionName.SKIP_TESTS;
 
 /**

--- a/compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java
+++ b/compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java
@@ -22,6 +22,7 @@ import io.ballerina.projects.environment.ProjectEnvironment;
 import io.ballerina.projects.internal.DefaultDiagnosticResult;
 import io.ballerina.projects.internal.PackageDiagnostic;
 import io.ballerina.projects.internal.jballerina.JarWriter;
+import io.ballerina.projects.internal.model.Target;
 import io.ballerina.projects.util.ProjectUtils;
 import io.ballerina.tools.diagnostics.Diagnostic;
 import io.ballerina.tools.diagnostics.DiagnosticSeverity;
@@ -103,7 +104,13 @@ public class JBallerinaBackend extends CompilerBackend {
 
     public static JBallerinaBackend from(PackageCompilation packageCompilation, JvmTarget jdkVersion) {
         // Check if the project has write permissions
-        checkWritePermission(packageCompilation.packageContext().project().sourceRoot());
+        if (packageCompilation.packageContext().project().kind().equals(ProjectKind.BUILD_PROJECT)) {
+            try {
+                new Target(packageCompilation.packageContext().project().sourceRoot());
+            } catch (IOException e) {
+                throw new ProjectException("error while checking permissions of target directory", e);
+            }
+        }
         return packageCompilation.getCompilerBackend(jdkVersion,
                 (targetPlatform -> new JBallerinaBackend(packageCompilation, jdkVersion)));
     }

--- a/compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/ProjectFiles.java
+++ b/compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/ProjectFiles.java
@@ -119,7 +119,11 @@ public class ProjectFiles {
     }
 
     public static List<DocumentData> loadDocuments(Path dirPath) {
-        checkReadPermission(dirPath);
+        try {
+            checkReadPermission(dirPath);
+        } catch (UnsupportedOperationException ignore) {
+            // ignore for zip entries
+        }
         try (Stream<Path> pathStream = Files.walk(dirPath, 1)) {
             return pathStream
                     .filter(BAL_EXTENSION_MATCHER::matches)
@@ -131,7 +135,11 @@ public class ProjectFiles {
     }
 
     private static List<DocumentData> loadTestDocuments(Path dirPath) {
-        checkReadPermission(dirPath);
+        try {
+            checkReadPermission(dirPath);
+        } catch (UnsupportedOperationException ignore) {
+            // ignore for zip entries
+        }
         try (Stream<Path> pathStream = Files.walk(dirPath, 1)) {
             return pathStream
                     .filter(BAL_EXTENSION_MATCHER::matches)
@@ -146,7 +154,11 @@ public class ProjectFiles {
         if (Files.notExists(documentFilePath)) {
             return null;
         }
-        checkReadPermission(documentFilePath);
+        try {
+            checkReadPermission(documentFilePath);
+        } catch (UnsupportedOperationException ignore) {
+            // ignore for zip entries
+        }
 
         String content;
         try {
@@ -158,7 +170,11 @@ public class ProjectFiles {
     }
 
     private static DocumentData loadTestDocument(Path documentFilePath) {
-        checkReadPermission(documentFilePath);
+        try {
+            checkReadPermission(documentFilePath);
+        } catch (UnsupportedOperationException ignore) {
+            // ignore for zip entries
+        }
         String content;
         try {
             content = Files.readString(documentFilePath, Charset.defaultCharset());

--- a/compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/ProjectFiles.java
+++ b/compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/ProjectFiles.java
@@ -119,6 +119,7 @@ public class ProjectFiles {
     }
 
     public static List<DocumentData> loadDocuments(Path dirPath) {
+        checkReadPermission(dirPath);
         try (Stream<Path> pathStream = Files.walk(dirPath, 1)) {
             return pathStream
                     .filter(BAL_EXTENSION_MATCHER::matches)
@@ -130,6 +131,7 @@ public class ProjectFiles {
     }
 
     private static List<DocumentData> loadTestDocuments(Path dirPath) {
+        checkReadPermission(dirPath);
         try (Stream<Path> pathStream = Files.walk(dirPath, 1)) {
             return pathStream
                     .filter(BAL_EXTENSION_MATCHER::matches)
@@ -144,6 +146,7 @@ public class ProjectFiles {
         if (Files.notExists(documentFilePath)) {
             return null;
         }
+        checkReadPermission(documentFilePath);
 
         String content;
         try {
@@ -155,6 +158,7 @@ public class ProjectFiles {
     }
 
     private static DocumentData loadTestDocument(Path documentFilePath) {
+        checkReadPermission(documentFilePath);
         String content;
         try {
             content = Files.readString(documentFilePath, Charset.defaultCharset());

--- a/compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/environment/BallerinaUserHome.java
+++ b/compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/environment/BallerinaUserHome.java
@@ -37,7 +37,8 @@ public final class BallerinaUserHome {
         try {
             Files.createDirectories(remotePackageRepositoryPath);
         } catch (IOException exception) {
-            throw new ProjectException("unable to create the file system cache of Ballerina Central repository");
+            throw new ProjectException("unable to create the file system cache of Ballerina Central repository: " +
+                    remotePackageRepositoryPath);
         }
 
         this.remotePackageRepository = RemotePackageRepository

--- a/compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/model/Target.java
+++ b/compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/model/Target.java
@@ -44,8 +44,6 @@ public class Target {
     private Path docPath;
 
     public Target(Path sourceRoot) throws IOException {
-        ProjectUtils.checkWritePermission(sourceRoot);
-
         this.targetPath = sourceRoot.resolve(ProjectConstants.TARGET_DIR_NAME);
         this.cache = this.targetPath.resolve(ProjectConstants.CACHES_DIR_NAME);
         this.balaCachePath = this.targetPath.resolve(ProjectConstants.TARGET_BALA_DIR_NAME);
@@ -55,7 +53,29 @@ public class Target {
         this.binPath = this.targetPath.resolve(ProjectConstants.BIN_DIR_NAME);
         this.reportPath = this.targetPath.resolve(ProjectConstants.REPORT_DIR_NAME);
         this.docPath = this.targetPath.resolve(ProjectConstants.TARGET_API_DOC_DIRECTORY);
-        Files.createDirectories(this.targetPath);
+
+        if (Files.exists(this.targetPath)) {
+            ProjectUtils.checkWritePermission(this.targetPath);
+        } else {
+            Files.createDirectories(this.targetPath);
+        }
+
+        if (Files.exists(this.cache)) {
+            ProjectUtils.checkWritePermission(this.cache);
+        }
+        if (Files.exists(this.binPath)) {
+            ProjectUtils.checkWritePermission(this.binPath);
+        }
+        if (Files.exists(this.balaCachePath)) {
+            ProjectUtils.checkWritePermission(this.balaCachePath);
+        }
+        if (Files.exists(this.docPath)) {
+            ProjectUtils.checkWritePermission(this.docPath);
+        }
+
+        if (Files.exists(this.reportPath)) {
+            ProjectUtils.checkWritePermission(this.reportPath);
+        }
     }
 
     /**
@@ -162,11 +182,15 @@ public class Target {
      */
     public void setOutputPath(Path outputPath) throws IOException {
         if (Files.exists(outputPath)) {
+            ProjectUtils.checkWritePermission(outputPath);
             Files.delete(outputPath);
         }
         // create parent directories
         Path parent = outputPath.getParent();
         if (parent != null) {
+            if (Files.exists(parent)) {
+                ProjectUtils.checkWritePermission(parent);
+            }
             Files.createDirectories(parent);
         }
         this.outputPath = outputPath;

--- a/compiler/ballerina-lang/src/main/java/io/ballerina/projects/util/ProjectUtils.java
+++ b/compiler/ballerina-lang/src/main/java/io/ballerina/projects/util/ProjectUtils.java
@@ -514,19 +514,19 @@ public class ProjectUtils {
 
     public static void checkWritePermission(Path path) {
         if (!path.toFile().canWrite()) {
-            throw new ProjectException("'" + path + "' does not have write permissions");
+            throw new ProjectException("'" + path.normalize() + "' does not have write permissions");
         }
     }
 
     public static void checkReadPermission(Path path) {
         if (!path.toFile().canRead()) {
-            throw new ProjectException("'" + path + "' does not have read permissions");
+            throw new ProjectException("'" + path.normalize() + "' does not have read permissions");
         }
     }
 
     public static void checkExecutePermission(Path path) {
         if (!path.toFile().canRead()) {
-            throw new ProjectException("'" + path + "' does not have execute permissions");
+            throw new ProjectException("'" + path.normalize() + "' does not have execute permissions");
         }
     }
 

